server/src/routes/bubble_guppies/jobs.rs

```diff
@@ -0,0 +1,318 @@
+    let peers_serialized = serde_json::to_string(&peers).expect("to serialize");
+
+    Response::builder()
+        .status(StatusCode::OK)
+        .body(Body::from(format!(r#"{{"peers":{peers_serialized}}}"#)))
+        .unwrap()
+}
+
+async fn get_job_list(State(state): State<ServerState>) -> impl IntoResponse {
+    let jobs = state.jobs.lock().await;
+    let jobs_list = jobs.get_jobs_list().await;
+
+    let jobs_json = serde_json::to_string(&jobs_list).expect("to serialize");
+    Response::builder()
+        .status(StatusCode::OK)
+        .body(Body::from(format!(r#"{{"jobs": {jobs_json}}}"#)))
+        .unwrap()
+}
+
+async fn get_job_info(
+    State(state): State<ServerState>,
+    Path(jobID): Path<String>,
+) -> impl IntoResponse {
+    let jobs = state.jobs.lock().await;
+
+    let job_info = match jobs.get_job_info(&jobID).await {
+        Some(job_info) => job_info,
+        None => return (StatusCode::NOT_FOUND, "Job not found").into_response(),
+    };
+
+    let info_json = serde_json::to_string(&job_info).unwrap();
+    Response::builder()
+        .status(StatusCode::OK)
+        .body(Body::from(info_json))
+        .unwrap()
+}
+
+#[derive(Debug, Serialize)]
+#[allow(non_snake_case)]
+struct JobPeerInfo {
+    ipAddress: String,
+    region: String,
+    name: String,
+    price: i64,
+}
+async fn job_peer_info(
+    State(state): State<ServerState>,
+    Path(jobID): Path<String>,
+) -> impl IntoResponse {
+    let jobs = state.jobs.lock().await;
+
+    let user = match jobs.get_job(&jobID).await {
+        Some(job) => {
+            let lock = job.lock().await;
+            match try_decode_user(lock.encoded_producer.as_str()) {
+                Ok(user) => user,
+                Err(_) => return (StatusCode::NOT_FOUND, "Failed to decode user").into_response(),
+            }
+        }
+        None => return (StatusCode::NOT_FOUND, "Job not found").into_response(),
+    };
+
+    let peer_info = JobPeerInfo {
+        ipAddress: user.ip,
+        region: "US".into(),
+        name: user.name,
+        price: user.price,
+    };
+
+    let info_json = serde_json::to_string(&peer_info).unwrap();
+    Response::builder()
+        .status(StatusCode::OK)
+        .body(Body::from(info_json))
+        .unwrap()
+}
+
+#[derive(Deserialize)]
+#[allow(non_snake_case)]
+struct JobIds {
+    jobIDs: Vec<String>,
+}
+async fn start_jobs(
+    State(state): State<ServerState>,
+    Json(arg): Json<JobIds>,
+) -> impl IntoResponse {
+    for job_id in arg.jobIDs {
+        let mut config = state.config.lock().await;
+        let jobs = state.jobs.lock().await;
+
+        match jobs.get_job(&job_id).await {
+            Some(job) => {
+                let token = config.get_token(job.lock().await.encoded_producer.clone());
+
+                jobs::start(job, state.jobs.clone(), token).await;
+            }
+            None => {
+                return (StatusCode::NOT_FOUND, format!("Job {job_id} not found")).into_response()
```

Comment:
Is start_jobs intended to end prematurely if one of the jobs is not found? pause_jobs and terminate_jobs is counting failures and will look at all the jobs even if one fails.

Reply:
Nope, that should be corrected.